packages/flutter/lib/src/widgets/basic.dart

```diff
@@ -397,15 +396,14 @@ class CustomOneChildLayout extends OneChildRenderObjectWidget {
   /// Whenever the delegate changes, we need to recompute the layout of this
   /// widget, which means you might not want to create a new delegate instance
   /// every time you build this widget. Instead, consider using a long-lived
-  /// deletate (perhaps held in a component's state) that you re-use every time
+  /// delegate (perhaps held in a component's state) that you re-use every time
   /// you build this widget.
   final OneChildLayoutDelegate delegate;
-  final Object token;
 
   RenderCustomOneChildLayoutBox createRenderObject() => new RenderCustomOneChildLayoutBox(delegate: delegate);
 
   void updateRenderObject(RenderCustomOneChildLayoutBox renderObject, CustomOneChildLayout oldWidget) {
-    if (oldWidget.token != token)
+    if (delegate == oldWidget.delegate && delegate.shouldRelayout())
       renderObject.markNeedsLayout();
```

Comment:
This work should be done in the render object, just like for RenderCustomPaint.  Notice that RenderCustomPaint also checks the runtimeType to make sure that shouldRelayout is always called with an argument that matches its runtimeType.